### ISSUE TITLE
lambda - init.js won`t be loaded correctly

### ISSUE THREAD

**Issue by pharindoko** [scottie1984/swagger-ui-express#177]
Hey Scottie,

nice work you`ve done with this tool.

I hope you can help me.
I`m using your solution in an aws lambda function. 
it proxies it to an express server (proxy integration). 

See here:
https://gt2l431zo1.execute-api.eu-central-1.amazonaws.com/dev/ui

**Please open chrome dev tools first and see what happens.**
Might be you need to load it a few times and you can see the swagger ui

The issue I do have is that the swagger-ui-init.js is missing for the first few lambdas so at startup - it responses 200 but has just a few bytes but not the code that is needed. 
after a few reloads it works...
https://gt2l431zo1.execute-api.eu-central-1.amazonaws.com/dev/swagger-ui-init.js

this is what I currently tried in code
```bash
    function: protected async setupSwagger ....
    //setup api endpoints
    ....
    CoreApp.swaggerSpec = this.apispec.generateSpecification(db, true);
    CoreApp.swaggerMiddleware = swaggerUi.setup(CoreApp.swaggerSpec);
    ...
    ...
    this.server.get('/ui/index.html', CoreApp.swaggerMiddleware);
    this.server.use('/ui', swaggerUi.serveWithOptions({ redirect: false }));

   // I do need this otherwise /ui without the trailing slash is not working
    this.server.use('/', swaggerUi.serveWithOptions({ redirect: false }));
    this.server.get('/ui', CoreApp.swaggerMiddleware);

```

any idea what could be the issue ?

br,

Florian

**Comment by scottie1984** [comment]
Do you have an example repository? My first guess is that the setupSwagger function you have created seems to be async and perhaps you are awaiting the spec correctly from the DB.

**Comment by pharindoko** [comment]
thanks for the reply - might be that I found the issue. 
I‘ll check it tomorrow and let you know.

**Comment by pharindoko** [comment]
here is the function containing the code.
https://github.com/pharindoko/json-serverless/blob/f41372f8cbcde26c43580b41441c8d3253b8a841/packages/server/src/app/core.app.ts#L88

I didn`t refactor it already as I`m still fighting with this :)


if you want to start the solution locally 
1. git clone
2. make start

Locally and "Serverless-Offline" it works.
1. In template folder there`s the serverless.yml containing serverless-offline.(check the makefile again "make start-offline")

**Comment by scottie1984** [comment]
My guess is :

```
generateSpecification = (json: object, regenerate: boolean) => {
    if (!this.spec || regenerate) {
```

That has something to do with the way the spec is generated as it is synchronously generated (I would recommend changing this to an async mechanism) and potentially this.spec is at some point left undefined. What I would do is manually set this.spec to some sample swagger doc and see do you still have the problem. If you don't then you can look into the way the swagger document is created. Also add some logging in at various points to trace through the problem. 

I would be particularly looking at what are the variables at the point swaggerUi.serveFiles is being called with.

Does   generateSpecification get called multiple times?

**Comment by pharindoko** [comment]
> My guess is :
> 
> ```
> generateSpecification = (json: object, regenerate: boolean) => {
>     if (!this.spec || regenerate) {
> ```
> 
> That has something to do with the way the spec is generated as it is synchronously generated (I would recommend changing this to an async mechanism) and potentially this.spec is at some point left undefined. What I would do is manually set this.spec to some sample swagger doc and see do you still have the problem. If you don't then you can look into the way the swagger document is created. Also add some logging in at various points to trace through the problem.
> 
> I would be particularly looking at what are the variables at the point swaggerUi.serveFiles is being called with.
> 
> Does generateSpecification get called multiple times?

Thanks a lot. It was a good idea to ask you.
I will check it the next few days. 

br,

Flo

**Comment by pharindoko** [comment]
![image](https://user-images.githubusercontent.com/5619511/80254043-b93aec00-867a-11ea-9a48-6a5073855e2e.png)

ok had a look - with a sample spec even the first request works without issues. So ok the swaggerspec doesn`t get generated at the moment I try to call it.

**Comment by rfarris** [comment]
This might be the same issue I'm seeing here:  #178

**Comment by pharindoko** [comment]
> This might be the same issue I'm seeing here: #178

Thanks for that hint. I will try this workaround.

**Comment by pharindoko** [comment]
closed this due to same issue describe in #178 
@scottie1984  would be nice if you can fix this race condition.
